src/algorithms/synthesis/syrec_synthesis.cpp

```diff
@@ -0,0 +1,1447 @@
+                if (i == j) {
+                    exp_op_vector.clear();
+                    exp_lhs_vector.clear();
+                    exp_rhs_vector.clear();
+                    return true;
+                }
+            }
+        }
+
+        exp_op_vector.clear();
+        exp_lhs_vector.clear();
+        exp_rhs_vector.clear();
+        return false;
+    }
+
+    /// generating LHS and RHS (not whole expressions, just the corresponding variables)
+    bool standard_syrec_synthesizer::op_rhs_lhs_expression(const applications::expression::ptr& expression, std::vector<unsigned>& v) {
+        if (auto* exp = dynamic_cast<applications::binary_expression*>(expression.get())) {
+            return op_rhs_lhs_expression(*exp, v);
+        } else if (auto* exp_2 = dynamic_cast<applications::variable_expression*>(expression.get())) {
+            return op_rhs_lhs_expression(*exp_2, v);
+        } else {
+            return false;
+        }
+    }
+
+    bool standard_syrec_synthesizer::op_rhs_lhs_expression(const applications::variable_expression& expression, std::vector<unsigned>& v) {
+        return get_variables(expression.var(), v);
+    }
+
+    bool standard_syrec_synthesizer::op_rhs_lhs_expression(const applications::binary_expression& expression, std::vector<unsigned>& v) {
+        std::vector<unsigned> lhs, rhs;
+
+        if (!op_rhs_lhs_expression(expression.lhs(), lhs) || !op_rhs_lhs_expression(expression.rhs(), rhs)) {
+            return false;
+        }
+
+        v = rhs;
+        op_vec.push_back(expression.op());
+        return true;
+    }
+
+    /////////When the input signals are not repeated//////////////////
+    bool standard_syrec_synthesizer::on_statement(const applications::statement::ptr& statement) {
+        _stmts.push(statement);
+        bool okay = false;
+        if (auto* stat = dynamic_cast<applications::swap_statement*>(statement.get())) {
+            okay = on_statement(*stat);
+        } else if (auto* stat_1 = dynamic_cast<applications::unary_statement*>(statement.get())) {
+            okay = on_statement(*stat_1);
+        } else if (auto* stat_2 = dynamic_cast<applications::assign_statement*>(statement.get())) {
+            okay = on_statement(*stat_2);
+        } else if (auto* stat_3 = dynamic_cast<applications::if_statement*>(statement.get())) {
+            okay = on_statement(*stat_3);
+        } else if (auto* stat_4 = dynamic_cast<applications::for_statement*>(statement.get())) {
+            okay = on_statement(*stat_4);
+        } else if (auto* stat_5 = dynamic_cast<applications::call_statement*>(statement.get())) {
+            okay = on_statement(*stat_5);
+        } else if (auto* stat_6 = dynamic_cast<applications::uncall_statement*>(statement.get())) {
+            okay = on_statement(*stat_6);
+        } else if (auto* stat_7 = dynamic_cast<applications::skip_statement*>(statement.get())) {
+            okay = on_statement(*stat_7);
```

Comment:
try to use concise naming. these `stat_x` variables can all have reasonable names, e.g., `swap`, `unary`, etc.